File: src/main/scala/com/yogpc/qp/packet/laser/LaserMessage.java
package com.yogpc.qp.packet.laser;

import java.util.stream.Stream;

import com.yogpc.qp.QuarryPlus;
import com.yogpc.qp.packet.IMessage;
import com.yogpc.qp.tile.TileLaser;
import net.minecraft.client.Minecraft;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

/**
 * To client only.
 */
public class LaserMessage implements IMessage {

    BlockPos pos;
    Vec3d[] vec3ds;

    public static LaserMessage create(TileLaser laser) {
        LaserMessage message = new LaserMessage();
        message.pos = laser.getPos();
        message.vec3ds = laser.lasers;
        return message;
    }

    @Override
    public void fromBytes(PacketBuffer buffer) {
        pos = buffer.readBlockPos();
        vec3ds = new Vec3d[buffer.readInt()];
        for (int i = 0; i < vec3ds.length; i++) {
            vec3ds[i] = new Vec3d(buffer.readDouble(), buffer.readDouble(), buffer.readDouble());
        }
    }

    @Override
    public void toBytes(PacketBuffer buffer) {
        buffer.writeBlockPos(pos);
        buffer.writeInt((int) Stream.of(vec3ds).filter(TileLaser.nonNull).count());
        Stream.of(vec3ds).filter(TileLaser.nonNull).forEach(vec3d -> buffer.writeDouble(vec3d.x).writeDouble(vec3d.y).writeDouble(vec3d.z));
    }

    @Override
    @SideOnly(Side.CLIENT)
    public IMessage onRecieve(IMessage message, MessageContext ctx) {
        World world = QuarryPlus.proxy.getPacketWorld(ctx.netHandler);
        TileLaser laser = (TileLaser) world.getTileEntity(pos);
        if (laser != null) {
            Minecraft.getMinecraft().addScheduledTask(() -> laser.lasers = vec3ds);
        }
        return null;
    }
}

File: src/main/scala/com/yogpc/qp/packet/laser/LaserMessage.scala
package com.yogpc.qp.packet.laser

import java.io.IOException
import java.util.Objects

import com.yogpc.qp.QuarryPlus
import com.yogpc.qp.packet.IMessage
import com.yogpc.qp.tile.TileLaser
import net.minecraft.client.Minecraft
import net.minecraft.network.PacketBuffer
import net.minecraft.util.math.{BlockPos, Vec3d}
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

object LaserMessage {
    def create(laser: TileLaser): LaserMessage = {
        val message = new LaserMessage(laser)
        message
    }
}

/**
  * To client only.
  */
class LaserMessage(laser: TileLaser) extends IMessage {
    private[this] var pos = BlockPos.ORIGIN
    private[this] var vec3ds: Array[Vec3d] = _

    def this() {
        this(null)
    }

    if (laser != null) {
        pos = laser.getPos
        vec3ds = laser.lasers
    }

    @throws[IOException]
    override def fromBytes(buffer: PacketBuffer): Unit = {
        pos = buffer.readBlockPos
        vec3ds = new Array[Vec3d](buffer.readInt)
        for (i <- vec3ds.indices) {
            vec3ds(i) = new Vec3d(buffer.readDouble, buffer.readDouble, buffer.readDouble)
        }
    }

    override def toBytes(buffer: PacketBuffer): Unit = {
        buffer.writeBlockPos(pos)
        buffer.writeInt(vec3ds.count(Objects.nonNull))
        for (vec3d <- vec3ds if vec3d != null) {
            buffer.writeDouble(vec3d.x).writeDouble(vec3d.y).writeDouble(vec3d.z)
        }
    }

    @SideOnly(Side.CLIENT) override def onRecieve(message: IMessage, ctx: MessageContext): IMessage = {
        val world = QuarryPlus.proxy.getPacketWorld(ctx.netHandler)
        val laser = world.getTileEntity(pos).asInstanceOf[TileLaser]
        if (laser != null) Minecraft.getMinecraft.addScheduledTask(new Runnable {
            override def run(): Unit = laser.lasers = vec3ds
        })
        null
    }
}
